File: src/com/dafttech/terra/game/world/Vector2i.java
package com.dafttech.terra.game.world;

import static com.dafttech.terra.resources.Options.BLOCK_SIZE;

import com.badlogic.gdx.Gdx;
import com.dafttech.terra.engine.Vector2;
import com.dafttech.terra.game.world.entities.Entity;

public class Vector2i {
    public int x, y; // TODO make private

    public Vector2i(int x, int y) {
        set(x, y);
    }

    public Vector2i(Vector2i pos) {
        set(pos);
    }

    public Vector2i(Vector2 pos) {
        set(pos);
    }

    public Vector2i() {
        set();
    }

    public Vector2 toScreenPos(Entity pointOfView) {
        return new Vector2(x * BLOCK_SIZE - pointOfView.getPosition().x + Gdx.graphics.getWidth() / 2, y * BLOCK_SIZE - pointOfView.getPosition().y
                + Gdx.graphics.getHeight() / 2);
    }

    public Vector2 toEntityPos() {
        return new Vector2(x * BLOCK_SIZE, y * BLOCK_SIZE);
    }

    public Vector2 toVector2() {
        return new Vector2(x, y);
    }

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    public Vector2i setX(int x) {
        this.x = x;
        return this;
    }

    public Vector2i setY(int y) {
        this.y = y;
        return this;
    }

    public Vector2i set() {
        this.x = 0;
        this.y = 0;
        return this;
    }

    public Vector2i set(int val) {
        this.x = val;
        this.y = val;
        return this;
    }

    public Vector2i set(int x, int y) {
        this.x = x;
        this.y = y;
        return this;
    }

    public Vector2i set(Vector2i pos) {
        this.x = pos.x;
        this.y = pos.y;
        return this;
    }

    public Vector2i set(Vector2 pos) {
        return set(pos.toWorldPosition());
        /*
         * this.x = (int) pos.x; this.y = (int) pos.y; return this;
         */
    }

    public Vector2i addX(int x) {
        this.x += x;
        return this;
    }

    public Vector2i addY(int y) {
        this.y += y;
        return this;
    }

    public Vector2i add() {
        this.x += 0;
        this.y += 0;
        return this;
    }

    public Vector2i add(int val) {
        this.x += val;
        this.y += val;
        return this;
    }

    public Vector2i add(int x, int y) {
        this.x += x;
        this.y += y;
        return this;
    }

    public Vector2i add(Vector2i pos) {
        this.x += pos.x;
        this.y += pos.y;
        return this;
    }

    public Vector2i add(Facing facing) {
        return add(facing.xOff, facing.yOff);
    }

    public Vector2i add(Vector2 pos) {
        return add(pos.toWorldPosition());
    }

    public Vector2i mulX(int x) {
        this.x *= x;
        return this;
    }

    public Vector2i mulY(int y) {
        this.y *= y;
        return this;
    }

    public Vector2i mul() {
        this.x *= 1;
        this.y *= 1;
        return this;
    }

    public Vector2i mul(int val) {
        this.x *= val;
        this.y *= val;
        return this;
    }

    public Vector2i mul(int x, int y) {
        this.x *= x;
        this.y *= y;
        return this;
    }

    public Vector2i mul(Vector2i pos) {
        this.x *= pos.x;
        this.y *= pos.y;
        return this;
    }

    public Vector2i mul(Vector2 pos) {
        return mul(pos.toWorldPosition());
    }

    public Vector2i setXNew(int x) {
        return clone().setX(x);
    }

    public Vector2i setYNew(int y) {
        return clone().setY(y);
    }

    public Vector2i setNew() {
        return clone().set();
    }

    public Vector2i setNew(int val) {
        return clone().setNew(val);
    }

    public Vector2i setNew(int x, int y) {
        return clone().set(x, y);
    }

    public Vector2i setNew(Vector2i pos) {
        return clone().set(pos);
    }

    public Vector2i setNew(Vector2 pos) {
        return clone().set(pos);
    }

    public Vector2i addXNew(int x) {
        return clone().addX(x);
    }

    public Vector2i addYNew(int y) {
        return clone().addY(y);
    }

    public Vector2i addNew() {
        return clone().add();
    }

    public Vector2i addNew(int val) {
        return clone().add(val);
    }

    public Vector2i addNew(int x, int y) {
        return clone().add(x, y);
    }

    public Vector2i addNew(Vector2i pos) {
        return clone().add(pos);
    }

    public Vector2i addNew(Vector2 pos) {
        return clone().add(pos);
    }

    public Vector2i addNew(Facing facing) {
        return addNew(facing.xOff, facing.yOff);
    }

    public Vector2i mulXNew(int x) {
        return clone().mulX(x);
    }

    public Vector2i mulYNew(int y) {
        return clone().mulY(y);
    }

    public Vector2i mulNew() {
        return clone().mul();
    }

    public Vector2i mulNew(int val) {
        return clone().mul(val);
    }

    public Vector2i mulNew(int x, int y) {
        return clone().mul(x, y);
    }

    public Vector2i mulNew(Vector2i pos) {
        return clone().mul(pos);
    }

    public Vector2i mulNew(Vector2 pos) {
        return clone().mul(pos);
    }

    @Override
    public Vector2i clone() {
        return new Vector2i(this);
    }

    @Override
    public boolean equals(Object obj) {
        return (obj instanceof Vector2i && ((Vector2i) obj).x == x && ((Vector2i) obj).y == y);
    }

    @Override
    public int hashCode() {
        return ((Integer) x).hashCode() + ((Integer) y).hashCode();
    }

    public boolean isInRect(int x, int y, int sizeX, int sizeY) {
        return this.x >= x && this.y >= y && this.x < x + sizeX && this.y < y + sizeY;
    }

    private static int getChunkPos(int blockInWorldPos, int chunkSize) {
        return blockInWorldPos >= 0 ? blockInWorldPos / chunkSize : (blockInWorldPos + 1) / chunkSize - 1;
    }

    private static int getBlockInChunkPos(int blockInWorldPos, int chunkSize) {
        return blockInWorldPos - getChunkPos(blockInWorldPos, chunkSize) * chunkSize;
    }

    private static int getBlockInWorldPos(int chunkPos, int chunkSize, int blockInChunkPos) {
        return chunkSize * chunkPos + blockInChunkPos;
    }

    public Vector2i getChunkPos(World world) {
        return new Vector2i(getChunkPos(x, world.chunksize.x), getChunkPos(y, world.chunksize.y));
    }

    public Vector2i getBlockInChunkPos(World world) {
        return new Vector2i(getBlockInChunkPos(x, world.chunksize.x), getBlockInChunkPos(y, world.chunksize.y));
    }

    public Vector2i getBlockInWorldPos(Chunk chunk) {
        return new Vector2i(getBlockInWorldPos(chunk.pos.x, chunk.world.chunksize.x, x), getBlockInWorldPos(chunk.pos.y, chunk.world.chunksize.y, y));
    }
}

File: src/main/scala/com/dafttech/terra/game/world/Vector2i.scala
package com.dafttech.terra.game.world

import com.badlogic.gdx.Gdx
import com.dafttech.terra.engine.Vector2
import com.dafttech.terra.game.world.entities.Entity
import com.dafttech.terra.resources.Options.BLOCK_SIZE

object Vector2i {
  private def getChunkPos(blockInWorldPos: Int, chunkSize: Int): Int = {
    return if (blockInWorldPos >= 0) blockInWorldPos / chunkSize else (blockInWorldPos + 1) / chunkSize - 1
  }

  private def getBlockInChunkPos(blockInWorldPos: Int, chunkSize: Int): Int = {
    return blockInWorldPos - getChunkPos(blockInWorldPos, chunkSize) * chunkSize
  }

  private def getBlockInWorldPos(chunkPos: Int, chunkSize: Int, blockInChunkPos: Int): Int = {
    return chunkSize * chunkPos + blockInChunkPos
  }
}
class Vector2i {
  var x: Int = 0
  var y: Int = 0
  set

  def this(x: Int, y: Int) {
    this()
    set(x, y)
  }

  def this(pos: Vector2i) {
    this()
    set(pos)
  }

  def this(pos: Vector2) {
    this()
    set(pos)
  }

  def toScreenPos(pointOfView: Entity): Vector2 = {
    return new Vector2(x * BLOCK_SIZE - pointOfView.getPosition.x + Gdx.graphics.getWidth / 2, y * BLOCK_SIZE - pointOfView.getPosition.y + Gdx.graphics.getHeight / 2)
  }

  def toEntityPos: Vector2 = {
    return new Vector2(x * BLOCK_SIZE, y * BLOCK_SIZE)
  }

  def toVector2: Vector2 = {
    return new Vector2(x, y)
  }

  def getX: Int = {
    return x
  }

  def getY: Int = {
    return y
  }

  def setX(x: Int): Vector2i = {
    this.x = x
    return this
  }

  def setY(y: Int): Vector2i = {
    this.y = y
    return this
  }

  def set: Vector2i = {
    this.x = 0
    this.y = 0
    return this
  }

  def set(`val`: Int): Vector2i = {
    this.x = `val`
    this.y = `val`
    return this
  }

  def set(x: Int, y: Int): Vector2i = {
    this.x = x
    this.y = y
    return this
  }

  def set(pos: Vector2i): Vector2i = {
    this.x = pos.x
    this.y = pos.y
    return this
  }

  def set(pos: Vector2): Vector2i = {
    return set(pos.toWorldPosition)
  }

  def addX(x: Int): Vector2i = {
    this.x += x
    return this
  }

  def addY(y: Int): Vector2i = {
    this.y += y
    return this
  }

  def add: Vector2i = {
    this.x += 0
    this.y += 0
    return this
  }

  def add(`val`: Int): Vector2i = {
    this.x += `val`
    this.y += `val`
    return this
  }

  def add(x: Int, y: Int): Vector2i = {
    this.x += x
    this.y += y
    return this
  }

  def add(pos: Vector2i): Vector2i = {
    this.x += pos.x
    this.y += pos.y
    return this
  }

  def add(facing: Facing): Vector2i = {
    return add(facing.xOff, facing.yOff)
  }

  def add(pos: Vector2): Vector2i = {
    return add(pos.toWorldPosition)
  }

  def mulX(x: Int): Vector2i = {
    this.x *= x
    return this
  }

  def mulY(y: Int): Vector2i = {
    this.y *= y
    return this
  }

  def mul: Vector2i = {
    this.x *= 1
    this.y *= 1
    return this
  }

  def mul(`val`: Int): Vector2i = {
    this.x *= `val`
    this.y *= `val`
    return this
  }

  def mul(x: Int, y: Int): Vector2i = {
    this.x *= x
    this.y *= y
    return this
  }

  def mul(pos: Vector2i): Vector2i = {
    this.x *= pos.x
    this.y *= pos.y
    return this
  }

  def mul(pos: Vector2): Vector2i = {
    return mul(pos.toWorldPosition)
  }

  def setXNew(x: Int): Vector2i = {
    return clone.setX(x)
  }

  def setYNew(y: Int): Vector2i = {
    return clone.setY(y)
  }

  def setNew: Vector2i = {
    return clone.set
  }

  def setNew(`val`: Int): Vector2i = {
    return clone.setNew(`val`)
  }

  def setNew(x: Int, y: Int): Vector2i = {
    return clone.set(x, y)
  }

  def setNew(pos: Vector2i): Vector2i = {
    return clone.set(pos)
  }

  def setNew(pos: Vector2): Vector2i = {
    return clone.set(pos)
  }

  def addXNew(x: Int): Vector2i = {
    return clone.addX(x)
  }

  def addYNew(y: Int): Vector2i = {
    return clone.addY(y)
  }

  def addNew: Vector2i = {
    return clone.add
  }

  def addNew(`val`: Int): Vector2i = {
    return clone.add(`val`)
  }

  def addNew(x: Int, y: Int): Vector2i = {
    return clone.add(x, y)
  }

  def addNew(pos: Vector2i): Vector2i = {
    return clone.add(pos)
  }

  def addNew(pos: Vector2): Vector2i = {
    return clone.add(pos)
  }

  def addNew(facing: Facing): Vector2i = {
    return addNew(facing.xOff, facing.yOff)
  }

  def mulXNew(x: Int): Vector2i = {
    return clone.mulX(x)
  }

  def mulYNew(y: Int): Vector2i = {
    return clone.mulY(y)
  }

  def mulNew: Vector2i = {
    return clone.mul
  }

  def mulNew(`val`: Int): Vector2i = {
    return clone.mul(`val`)
  }

  def mulNew(x: Int, y: Int): Vector2i = {
    return clone.mul(x, y)
  }

  def mulNew(pos: Vector2i): Vector2i = {
    return clone.mul(pos)
  }

  def mulNew(pos: Vector2): Vector2i = {
    return clone.mul(pos)
  }

  override def clone: Vector2i = {
    return new Vector2i(this)
  }

  override def equals(obj: Any): Boolean = {
    return (obj.isInstanceOf[Vector2i] && (obj.asInstanceOf[Vector2i]).x == x && (obj.asInstanceOf[Vector2i]).y == y)
  }

  override def hashCode: Int = {
    return (x.asInstanceOf[Integer]).hashCode * 5573 + (y.asInstanceOf[Integer]).hashCode
  }

  def isInRect(x: Int, y: Int, sizeX: Int, sizeY: Int): Boolean = {
    return this.x >= x && this.y >= y && this.x < x + sizeX && this.y < y + sizeY
  }

  def getChunkPos(world: World): Vector2i = {
    return new Vector2i(Vector2i.getChunkPos(x, world.chunksize.x), Vector2i.getChunkPos(y, world.chunksize.y))
  }

  def getBlockInChunkPos(world: World): Vector2i = {
    return new Vector2i(Vector2i.getBlockInChunkPos(x, world.chunksize.x), Vector2i.getBlockInChunkPos(y, world.chunksize.y))
  }

  def getBlockInWorldPos(chunk: Chunk): Vector2i = {
    return new Vector2i(Vector2i.getBlockInWorldPos(chunk.pos.x, chunk.world.chunksize.x, x), Vector2i.getBlockInWorldPos(chunk.pos.y, chunk.world.chunksize.y, y))
  }
}